File: src/account_management/account.py
import asyncio
from solana.rpc.async_api import AsyncClient
from dotenv import dotenv_values
import solders
import base58
import time
import json

class Account:
    def __init__(self):
        self.pk = dotenv_values()["PRIVATE_KEY"]
        self.keypair = solders.keypair.Keypair.from_base58_string(self.pk)
        self.public_key = self.keypair.pubkey()
        
    async def sign_in(self, timestamp: int = None)-> dict:
        if timestamp is None:
            timestamp = int(time.time() * 1000)
            timestamp = 1713298780638
            to_sign = f"Sign in to pump.fun: {timestamp}"
        signature = self.keypair.sign_message(to_sign.encode())
        body = {"address":str(self.keypair.pubkey()), "signature":str(signature) , "timestamp": int(timestamp)}
        return body

File: node-scripts/account_management/account.ts
import * as dotenv from 'dotenv';
import { Keypair } from '@solana/web3.js';
import * as nacl from 'tweetnacl';
import axios from 'axios';

dotenv.config();

class Account {
    private keypair: Keypair;

    constructor() {
        const privateKey = process.env.PRIVATE_KEY;
        if (!privateKey) {
            throw new Error('Private key not found in environment variables');
        }
        const secretKey = Buffer.from(privateKey, 'base64');
        this.keypair = Keypair.fromSecretKey(secretKey);
    }

    public async signIn(timestamp: Date): Promise<any> {
        const message = new TextEncoder().encode(timestamp.toISOString());
        const signature = nacl.sign.detached(message, this.keypair.secretKey);
        const body = {
            address: this.keypair.publicKey.toBase58(),
            signature: Buffer.from(signature).toString('base64'),
            timestamp: timestamp.toISOString()
        };
        const headers = {
            'Content-Type': 'application/json'
            // Add any other necessary headers here
        };

        try {
            const response = await axios.post('https://client-api.devnet.pump.fun/auth/login', body, { headers });
            return response.data;
        } catch (error) {
            console.error('Error in signIn:', error);
            throw error;
        }
    }
}

async function main() {
    const account = new Account();
    const signInResult = await account.signIn(new Date());
    console.log('Sign In Result:', signInResult);
}

main();